File: schemas/subscription-membership/drizzle/invoices.ts
// invoices: Billing documents generated for customers, optionally tied to subscriptions.
// See README.md for full design rationale.

import {
  pgTable,
  pgEnum,
  uuid,
  text,
  integer,
  timestamp,
  jsonb,
  index,
} from "drizzle-orm/pg-core";
import { customers } from "./customers";
import { subscriptions } from "./subscriptions";

export const invoiceStatusEnum = pgEnum("invoice_status", [
  "draft",
  "open",
  "paid",
  "void",
  "uncollectible",
]);

export const invoices = pgTable(
  "invoices",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    customerId: uuid("customer_id")
      .notNull()
      .references(() => customers.id, { onDelete: "cascade" }),
    subscriptionId: uuid("subscription_id").references(() => subscriptions.id, { onDelete: "set null" }),
    status: invoiceStatusEnum("status").notNull().default("draft"),
    currency: text("currency").notNull(),
    subtotal: integer("subtotal").notNull().default(0),
    tax: integer("tax").notNull().default(0),
    total: integer("total").notNull().default(0),
    amountPaid: integer("amount_paid").notNull().default(0),
    amountDue: integer("amount_due").notNull().default(0),
    periodStart: timestamp("period_start", { withTimezone: true }),
    periodEnd: timestamp("period_end", { withTimezone: true }),
    dueDate: timestamp("due_date", { withTimezone: true }),
    paidAt: timestamp("paid_at", { withTimezone: true }),
    hostedInvoiceUrl: text("hosted_invoice_url"),
    invoicePdfUrl: text("invoice_pdf_url"),
    metadata: jsonb("metadata"),
    providerType: text("provider_type"),
    providerId: text("provider_id"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_invoices_customer_id").on(table.customerId),
    index("idx_invoices_subscription_id").on(table.subscriptionId),
    index("idx_invoices_status").on(table.status),
    index("idx_invoices_provider").on(table.providerType, table.providerId),
  ]
);

File: schemas/subscription-membership/spacetimedb/invoices.rs
// invoices: Billing documents generated for customers per billing cycle.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum InvoiceStatus {
    Draft,          // type: String
    Open,
    Paid,
    Void,
    Uncollectible,
}

#[spacetimedb::table(name = invoices, public)]
pub struct Invoice {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub customer_id: String, // UUID — FK → customers.id (cascade delete)
    pub subscription_id: Option<String>, // UUID — FK → subscriptions.id (set null)
    #[index(btree)]
    pub status: InvoiceStatus,
    pub currency: String,
    pub subtotal: i32,
    pub tax: i32,
    pub total: i32,
    pub amount_paid: i32,
    pub amount_due: i32,
    pub period_start: Option<Timestamp>,
    pub period_end: Option<Timestamp>,
    pub due_date: Option<Timestamp>,
    pub paid_at: Option<Timestamp>,
    pub hosted_invoice_url: Option<String>,
    pub invoice_pdf_url: Option<String>,
    pub metadata: Option<String>, // JSON
    pub provider_type: Option<String>,
    pub provider_id: Option<String>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
